Merge Robot Classes

BD103 opened the pull request:
_Whew_

This fixes #9, but should not be merged until it can be fully tested with the robot. We can discuss the code design, but I tried to add helpful comments along the way. Happy holidays! ⛄

BD103 committed:
feat: add armBase motors

BD103 committed:
fix: change armBase type to servo

BD103 committed:
feat: add more functionality to MoveController

Arm base still doesn't work, though.

BD103 committed:
feat: add claw

Also add some telemetry stuff

Commit:
Arm Base working

Commit:
Movement code

BD103 committed:
feat!: motor modifiers

Commit:
Update MoveController.java

Commit:
Merge branch 'first-scrimmage' of https://github.com/BotsBurgh/BOTSBURGH-FTC-2021-22 into first-scrimmage

BD103 committed:
feat: add safeguard against moving too far

BD103 committed:
chore: remove duplicate v2 subpackage

BD103 committed:
feat: add better controller system

Commit:
Merge branch 'first-scrimmage' of https://github.com/BotsBurgh/BOTSBURGH-FTC-2021-22 into first-scrimmage

Commit:
Arm and Claw bug changes

9p4 committed:
Fix all of the things

BD103 committed:
fix: slow down duck and shorten autonomous

Co-Authored-By: BenRoss00000 <90357934+BenRoss00000@users.noreply.github.com>

9p4 committed:
Update API to modernity

9p4 committed:
New API, fixed naming, modernize opmodes

9p4 committed:
Fix telemetry

9p4 committed:
Merge branch 'proper-auto' into first-scrimmage

9p4 committed:
Update depot

BD103 committed:
refactor: rename Robot to OldRobot

BD103 committed:
feat: new robot design

(Again XD)

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/InitRobot.java:
```diff
@@ -0,0 +1,139 @@
+ * The Robot Initializer. Place initialization code here. This prevents needing to sync the init code
+ * between all OpModes.
+ */
+public class InitRobot {
```
Shouldn't this be renamed to be relevant to the merging of the robot classes?

9p4 replied:
So the init should be a part of the robot class specific to the robot that this class is initializing

BD103 replied:
InitRobot is going to be deleted, replaced by Robot.initTeleOp method.

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/OldRobot.java:
```diff
@@ -0,0 +1,71 @@
+ * add it to our GitHub issues page at https://github.com/botsburgh/BOTSBURGH-FTC-2020-21/issues
+ */
+@Builder
+public class OldRobot {
```
Why are we calling this OldRobot?

BD103 replied:
This is so I don't delete old code until I can confirm that the new code works.

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java:
```diff
@@ -2,73 +2,150 @@
-        this.backLeft = opMode.hardwareMap.get(DcMotor.class, Hardware.BACK_LEFT);
-        this.backRight = opMode.hardwareMap.get(DcMotor.class, Hardware.BACK_RIGHT);
-        this.duck = opMode.hardwareMap.get(DcMotor.class, Hardware.DUCK);
+    public void powerWheels(double flPower, double frPower, double blPower, double brPower) {
```
Is this redundant because of the movement class?

BD103 replied:
The movement class is going to be deleted, so no.

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java:
```diff
@@ -2,73 +2,150 @@
-        this.frontRight.setPower(power);
-        this.backLeft.setPower(power);
-        this.backRight.setPower(power);
+    public void positionClaw(double position) {
```
Is this redundant because of the movement class?

BD103 replied:
See above comment.

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java:
```diff
@@ -2,73 +2,150 @@
-        this.backLeft.setPower(0);
-        this.backRight.setPower(0);
-        this.duck.setPower(0);
+    public void adjustClaw(double amount) {
```
Is this redundant because of the movement class?

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java:
```diff
@@ -2,73 +2,150 @@
-    public Robot() {
-        super();
-    }
+public class Robot extends AbstractRobot implements WheeledRobot, ArmRobot {
```
Most of the functions here are redundant (implemented in the movement class). Pick which ones you would want to keep

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java:
```diff
@@ -2,73 +2,150 @@
-    public void setMotorsDirection(@NonNull DcMotor[] motors, DcMotor.Direction direction) {
-        for (DcMotor motor: motors) {
-            motor.setDirection(direction);
+    @Override
```
Merge with InitRobot class

BD103 replied:
InitRobot will not work with a class-based robot configuration. Sense we are starting to have multiple versions of the robot, the new one might have sensors and motors that the old doesn't. If we initialize this hardware in a static class that cannot change, there will be errors. Nesting the initialization inside of the Robot class means that each robot can have its own hardware. Therefor, InitRobot is going to be deleted.

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/bp/SensorRobot.java:
```diff
@@ -0,0 +1,12 @@
+import org.firstinspires.ftc.teamcode.api.hw.SmartColorSensor.Color;
+
+public interface SensorRobot {
+    // Allow having a power default?
```
I would recommend this, yes. Overload it maybe?

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/bp/WheeledRobot.java:
```diff
@@ -0,0 +1,8 @@
+package org.firstinspires.ftc.teamcode.api.bp;
+
+public interface WheeledRobot {
```
Again, redundant by movement class. Pick which one you would like to keep

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/config/Constants.java:
```diff
@@ -0,0 +1,16 @@
+    public static final double PARK_GREEN_FUDGE = 150;
+    public static final double PARK_BLUE_FUDGE  = 150;
+
+    // Drive Adjustments
```
We'll need to do some testing to refine these values.

BD103 replied:
I created #12 for this, resolving.

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/autonomous/Depot.java:
```diff
@@ -0,0 +1,23 @@
+public class Depot extends LinearOpMode {
+    @Override
+    public void runOpMode() {
+        InitRobot.init(this);
```
Update to use new abstractions

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/ColorSensorTelemetry.java:
```diff
@@ -0,0 +1,64 @@
+public class ColorSensorTelemetry extends LinearOpMode {
+    @Override
+    public void runOpMode() throws InterruptedException {
+        InitRobot.init(this);
```
Update to use new abstractions

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/Forward.java:
```diff
@@ -0,0 +1,56 @@
+public class Forward extends LinearOpMode {
+    @Override
+    public void runOpMode() {
+        InitRobot.init(this);
```
Update to use new abstractions

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/opmode/teleop/TeleOpMain.java:
```diff
@@ -0,0 +1,93 @@
+public class TeleOpMain extends LinearOpMode {
+    @Override
+    public void runOpMode() {
+        InitRobot.init(this);
```
Update to use new abstractions

9p4 reviewed (changes requested):
So far so good. LMK when you've made these changes.

BD103 committed:
fix: make robot work with current robot configuration

* dying *

BD103 committed:
feat: make depot use merged robot

BD103 committed:
feat: convert all programs to new system

BD103 committed:
fix: direction of arm and claw

BD103 committed:
feat: have a power default for SensorRobot

https://github.com/BotsBurgh/BOTSBURGH-FTC-2021-22/pull/11#discussion_r779146492

9p4 commented on TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/Robot.java:
```diff
@@ -2,73 +2,157 @@
+        */
+
+        // Looped Config
+        SmartMotor[] wheelLoop = new SmartMotor[]{this.fl, this.fr, this.bl, this.br};
```
Perhaps use a hashmap for the motors instead so that we don't have to keep track of indexes and instead can use strings?

BD103 replied:
Hash maps are a pain to work with in Java. If it were Python, the I’d probably say yes. For Java, though, it seems better to use direct variables and avoid Nonnulls. Worst case scenario, we’ll just refactor in Android Studio.

9p4 commented:
Can you rebase so we don't have to worry about merge conflicts?

BD103 commented:
> Can you rebase so we don't have to worry about merge conflicts?

I can once it gets approved.

9p4 commented:
LGTM. Working on resolving conflicts

BD103 commented:
🦀 🦀 🦀
